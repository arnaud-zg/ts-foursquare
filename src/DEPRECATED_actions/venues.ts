import { ActionType, createAsyncAction } from 'typesafe-actions'
import { NRequest } from '../../types'
import {
  adaptGetVenuesCategories,
  adaptGetVenuesExplore,
  adaptGetVenuesLikes,
  adaptGetVenuesListed,
  adaptGetVenuesNextVenues,
  adaptGetVenuesSearch,
  adaptGetVenuesSimilar,
  adaptGetVenuesSuggestCompletion,
  adaptGetVenuesTrending,
} from '../adapter/venues'

export const getVenuesCategoriesAsync = createAsyncAction(
  'GET_VENUES_CATEGORIES_REQUEST',
  'GET_VENUES_CATEGORIES_SUCCESS',
  'GET_VENUES_CATEGORIES_FAILURE',
  'GET_VENUES_CATEGORIES_CANCEL'
)<void, ReturnType<typeof adaptGetVenuesCategories>, Error, string>()

export const getVenuesExploreAsync = createAsyncAction(
  'GET_VENUES_EXPLORE_REQUEST',
  'GET_VENUES_EXPLORE_SUCCESS',
  'GET_VENUES_EXPLORE_FAILURE',
  'GET_VENUES_EXPLORE_CANCEL'
)<
  NRequest.TVenuesExplorePayload,
  ReturnType<typeof adaptGetVenuesExplore>,
  Error,
  string
>()

export const getVenuesLikesAsync = createAsyncAction(
  'GET_VENUES_LIKES_REQUEST',
  'GET_VENUES_LIKES_SUCCESS',
  'GET_VENUES_LIKES_FAILURE',
  'GET_VENUES_LIKES_CANCEL'
)<
  NRequest.IVenuesLikesPayload,
  ReturnType<typeof adaptGetVenuesLikes>,
  Error,
  string
>()

export const getVenuesListedAsync = createAsyncAction(
  'GET_VENUES_LISTED_REQUEST',
  'GET_VENUES_LISTED_SUCCESS',
  'GET_VENUES_LISTED_FAILURE',
  'GET_VENUES_LISTED_CANCEL'
)<
  NRequest.IVenuesListedPayload,
  ReturnType<typeof adaptGetVenuesListed>,
  Error,
  string
>()

export const getVenuesNextVenuesAsync = createAsyncAction(
  'GET_VENUES_NEXT_VENUES_REQUEST',
  'GET_VENUES_NEXT_VENUES_SUCCESS',
  'GET_VENUES_NEXT_VENUES_FAILURE',
  'GET_VENUES_NEXT_VENUES_CANCEL'
)<
  NRequest.IVenuesNextVenuesPayload,
  ReturnType<typeof adaptGetVenuesNextVenues>,
  Error,
  string
>()

export const getVenuesSearchAsync = createAsyncAction(
  'GET_VENUES_SEARCH_REQUEST',
  'GET_VENUES_SEARCH_SUCCESS',
  'GET_VENUES_SEARCH_FAILURE',
  'GET_VENUES_SEARCH_CANCEL'
)<
  NRequest.TVenuesSearchPayload,
  ReturnType<typeof adaptGetVenuesSearch>,
  Error,
  string
>()

export const getVenuesSimilarAsync = createAsyncAction(
  'GET_VENUES_SIMILAR_REQUEST',
  'GET_VENUES_SIMILAR_SUCCESS',
  'GET_VENUES_SIMILAR_FAILURE',
  'GET_VENUES_SIMILAR_CANCEL'
)<
  NRequest.IVenuesSimilarPayload,
  ReturnType<typeof adaptGetVenuesSimilar>,
  Error,
  string
>()

export const getVenuesSuggestCompletionAsync = createAsyncAction(
  'GET_VENUES_SUGGEST_COMPLETION_REQUEST',
  'GET_VENUES_SUGGEST_COMPLETION_SUCCESS',
  'GET_VENUES_SUGGEST_COMPLETION_FAILURE',
  'GET_VENUES_SUGGEST_COMPLETION_CANCEL'
)<
  NRequest.TVenuesSuggestCompletionPayload,
  ReturnType<typeof adaptGetVenuesSuggestCompletion>,
  Error,
  string
>()

export const getVenuesTrendingAsync = createAsyncAction(
  'GET_VENUES_TRENDING_REQUEST',
  'GET_VENUES_TRENDING_SUCCESS',
  'GET_VENUES_TRENDING_FAILURE',
  'GET_VENUES_TRENDING_CANCEL'
)<
  NRequest.TVenuesTrendingPayload,
  ReturnType<typeof adaptGetVenuesTrending>,
  Error,
  string
>()

export const venuesActions = {
  getVenuesCategoriesAsync,
  getVenuesExploreAsync,
  getVenuesLikesAsync,
  getVenuesListedAsync,
  getVenuesNextVenuesAsync,
  getVenuesSearchAsync,
  getVenuesSimilarAsync,
  getVenuesSuggestCompletionAsync,
  getVenuesTrendingAsync,
}

export type TVenuesAction = ActionType<typeof venuesActions>
