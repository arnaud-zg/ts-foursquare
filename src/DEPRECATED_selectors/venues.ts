import { NStore, NVenue } from '../../types'

export const venueCategoriesStateSelector = (
  state: NStore.IState,
  venueId: string
): NVenue.ICategory[] => {
  const venueEntity = venuesEntitiesSelector(state)[venueId]
  return venueEntity ? venueEntity.categories : []
}

export const venueLocationSelector = (
  state: NStore.IState,
  venueId: string
): NVenue.ILocation | null => {
  const venueEntity = venuesEntitiesSelector(state)[venueId]
  return venueEntity ? venueEntity.location : null
}

export const venuesCategoriesSelector = (state: NStore.IState) =>
  venuesSelector(state).categories

export const venueSelector = (state: NStore.IState, venueId: string) =>
  venuesEntitiesSelector(state)[venueId]

export const venuesEntitiesSelector = (state: NStore.IState) =>
  venuesSelector(state).entities

export const venuesNextVenuesSelector = (state: NStore.IState) =>
  venuesSelector(state).nextVenues

export const venuesSimilarSelector = (state: NStore.IState) =>
  venuesSelector(state).similarVenues

export const venuesRecommendedPlacesSelector = (state: NStore.IState) =>
  venuesSelector(state).recommendedPlaces

export const venuesListedSelector = (state: NStore.IState) =>
  venuesSelector(state).venuesListed

export const venuesListedGroupSelector = (
  state: NStore.IState,
  groupKey: string
) => venuesSelector(state).venuesListed[groupKey]

export const venuesSelector = (state: NStore.IState) => state.venues
