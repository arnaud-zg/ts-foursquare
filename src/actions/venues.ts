import { ActionType, createAsyncAction } from 'typesafe-actions'
import {
  NLikes,
  NMiniVenue,
  NRecommendedPlaces,
  NRequest,
  NVenue,
  NVenuesCategories,
} from '../../types'

export const getVenuesCategoriesAsync = createAsyncAction(
  'GET_VENUES_CATEGORIES_REQUEST',
  'GET_VENUES_CATEGORIES_SUCCESS',
  'GET_VENUES_CATEGORIES_FAILURE',
  'GET_VENUES_CATEGORIES_CANCEL'
)<void, NVenuesCategories.ICategory[], Error, string>()

export const getVenuesExploreAsync = createAsyncAction(
  'GET_VENUES_EXPLORE_REQUEST',
  'GET_VENUES_EXPLORE_SUCCESS',
  'GET_VENUES_EXPLORE_FAILURE',
  'GET_VENUES_EXPLORE_CANCEL'
)<
  NRequest.TVenuesExplorePayload,
  NRecommendedPlaces.IGroupItem[],
  Error,
  string
>()

export const getVenuesLikesAsync = createAsyncAction(
  'GET_VENUES_LIKES_REQUEST',
  'GET_VENUES_LIKES_SUCCESS',
  'GET_VENUES_LIKES_FAILURE',
  'GET_VENUES_LIKES_CANCEL'
)<NRequest.IVenuesLikesPayload, NLikes.ILikes, Error, string>()

export const getVenuesNextVenuesAsync = createAsyncAction(
  'GET_VENUES_NEXT_VENUES_REQUEST',
  'GET_VENUES_NEXT_VENUES_SUCCESS',
  'GET_VENUES_NEXT_VENUES_FAILURE',
  'GET_VENUES_NEXT_VENUES_CANCEL'
)<NRequest.IVenuesNextVenuesPayload, NVenue.IVenue[], Error, string>()

export const getVenuesSearchAsync = createAsyncAction(
  'GET_VENUES_SEARCH_REQUEST',
  'GET_VENUES_SEARCH_SUCCESS',
  'GET_VENUES_SEARCH_FAILURE',
  'GET_VENUES_SEARCH_CANCEL'
)<NRequest.TVenuesSearchPayload, NVenue.IVenue[], Error, string>()

export const getVenuesSimilarAsync = createAsyncAction(
  'GET_VENUES_SIMILAR_REQUEST',
  'GET_VENUES_SIMILAR_SUCCESS',
  'GET_VENUES_SIMILAR_FAILURE',
  'GET_VENUES_SIMILAR_CANCEL'
)<NRequest.IVenuesSimilarPayload, NVenue.IVenue[], Error, string>()

export const getVenuesTrendingAsync = createAsyncAction(
  'GET_VENUES_TRENDING_REQUEST',
  'GET_VENUES_TRENDING_SUCCESS',
  'GET_VENUES_TRENDING_FAILURE',
  'GET_VENUES_TRENDING_CANCEL'
)<NRequest.TVenuesTrendingPayload, NVenue.IVenue[], Error, string>()

export const getVenuesSuggestCompletionAsync = createAsyncAction(
  'GET_VENUES_SUGGEST_COMPLETION_REQUEST',
  'GET_VENUES_SUGGEST_COMPLETION_SUCCESS',
  'GET_VENUES_SUGGEST_COMPLETION_FAILURE',
  'GET_VENUES_SUGGEST_COMPLETION_CANCEL'
)<
  NRequest.TVenuesSuggestCompletionPayload,
  NMiniVenue.IMiniVenue[],
  Error,
  string
>()

export const venuesActions = {
  getVenuesCategoriesAsync,
  getVenuesExploreAsync,
  getVenuesLikesAsync,
  getVenuesNextVenuesAsync,
  getVenuesSearchAsync,
  getVenuesSimilarAsync,
  getVenuesSuggestCompletionAsync,
  getVenuesTrendingAsync,
}

export type TVenuesAction = ActionType<typeof venuesActions>
