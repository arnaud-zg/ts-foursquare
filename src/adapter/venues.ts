import { NHttpStatuses } from 'http-response-status'
import { NPayload } from '../../types/payload.d'
import { NRecommendedPlaces } from '../../types/recommendedPlaces'
import { NVenue } from '../../types/venue.d'

export const adaptGetVenuesSearch = (
  payload: NPayload.IPayload<NVenue.IResponse>
): NVenue.IVenue[] =>
  !!payload.meta &&
  payload.meta.code === NHttpStatuses.ESuccess.OK &&
  !!payload.response &&
  !!payload.response.venues &&
  !!payload.response.venues.length
    ? payload.response.venues
    : []

export const adaptGetVenuesExplore = (
  payload: NPayload.IPayload<NRecommendedPlaces.IResponse>
): NRecommendedPlaces.IGroupItem[] =>
  !!payload.meta &&
  payload.meta.code === NHttpStatuses.ESuccess.OK &&
  !!payload.response &&
  !!payload.response.groups &&
  !!payload.response.groups.length &&
  !!payload.response.groups[0].items &&
  !!payload.response.groups[0].items.length
    ? payload.response.groups[0].items
    : []

export const adaptGetVenuesTrending = (
  payload: NPayload.IPayload<NVenue.IResponse>
): NVenue.IVenue[] => adaptGetVenuesSearch(payload)
