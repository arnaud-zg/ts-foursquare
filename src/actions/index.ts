import { TLifeAction } from './life'
import { TVenuesAction } from './venues'

export type TRootActionType = TVenuesAction | TLifeAction

export type TRootAction = TVenuesAction | TLifeAction

export * from './life'
export * from './venues'
