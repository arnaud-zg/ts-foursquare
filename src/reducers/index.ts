import { NVenues } from '../../types/venues'

export type TRootState = {
  venues: NVenues.IVenuesState
}
