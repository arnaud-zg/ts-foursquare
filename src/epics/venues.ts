import { Epic } from 'redux-observable'
import { of } from 'rxjs/internal/observable/of'
import { catchError, filter, map, switchMap, takeUntil } from 'rxjs/operators'
import { isActionOf } from 'typesafe-actions'
import { NStore } from '../../types/store'
import { TRootAction } from '../actions'
import {
  getVenuesCategoriesAsync,
  getVenuesExploreAsync,
  getVenuesLikesAsync,
  getVenuesNextVenuesAsync,
  getVenuesSearchAsync,
  getVenuesSimilarAsync,
  getVenuesSuggestCompletionAsync,
  getVenuesTrendingAsync,
} from '../actions/venues'
import {
  adaptGetVenuesCategories,
  adaptGetVenuesExplore,
  adaptGetVenuesLikes,
  adaptGetVenuesNextVenues,
  adaptGetVenuesSearch,
  adaptGetVenuesSimilar,
  adaptGetVenuesSuggestCompletion,
  adaptGetVenuesTrending,
} from '../adapter/venues'
import {
  getObservableVenuesCategories,
  getObservableVenuesExplore,
  getObservableVenuesLikes,
  getObservableVenuesNextVenues,
  getObservableVenuesSearch,
  getObservableVenuesSimilar,
  getObservableVenuesSuggestCompletion,
  getObservableVenuesTrending,
} from '../services/venues'

export const getVenuesCategoriesEpic: Epic<
  TRootAction,
  TRootAction,
  NStore.IState
> = (action$, state$) =>
  action$.pipe(
    filter(isActionOf(getVenuesCategoriesAsync.request)),
    switchMap(() =>
      getObservableVenuesCategories({ state$ }).pipe(
        map(adaptGetVenuesCategories),
        map(getVenuesCategoriesAsync.success),
        catchError(err => of(getVenuesCategoriesAsync.failure(err))),
        takeUntil(
          action$.pipe(filter(isActionOf(getVenuesCategoriesAsync.cancel)))
        )
      )
    )
  )

export const getVenuesExploreEpic: Epic<
  TRootAction,
  TRootAction,
  NStore.IState
> = (action$, state$) =>
  action$.pipe(
    filter(isActionOf(getVenuesExploreAsync.request)),
    switchMap(action =>
      getObservableVenuesExplore({ action, state$ }).pipe(
        map(adaptGetVenuesExplore),
        map(getVenuesExploreAsync.success),
        catchError(err => of(getVenuesExploreAsync.failure(err))),
        takeUntil(
          action$.pipe(filter(isActionOf(getVenuesExploreAsync.cancel)))
        )
      )
    )
  )

export const getVenuesLikesEpic: Epic<
  TRootAction,
  TRootAction,
  NStore.IState
> = (action$, state$) =>
  action$.pipe(
    filter(isActionOf(getVenuesLikesAsync.request)),
    switchMap(action =>
      getObservableVenuesLikes({ action, state$ }).pipe(
        map(adaptGetVenuesLikes),
        map(getVenuesLikesAsync.success),
        catchError(err => of(getVenuesLikesAsync.failure(err))),
        takeUntil(action$.pipe(filter(isActionOf(getVenuesLikesAsync.cancel))))
      )
    )
  )

export const getVenuesNextVenuesEpic: Epic<
  TRootAction,
  TRootAction,
  NStore.IState
> = (action$, state$) =>
  action$.pipe(
    filter(isActionOf(getVenuesNextVenuesAsync.request)),
    switchMap(action =>
      getObservableVenuesNextVenues({ action, state$ }).pipe(
        map(adaptGetVenuesNextVenues),
        map(getVenuesNextVenuesAsync.success),
        catchError(err => of(getVenuesNextVenuesAsync.failure(err))),
        takeUntil(
          action$.pipe(filter(isActionOf(getVenuesNextVenuesAsync.cancel)))
        )
      )
    )
  )

export const getVenuesSearchEpic: Epic<
  TRootAction,
  TRootAction,
  NStore.IState
> = (action$, state$) =>
  action$.pipe(
    filter(isActionOf(getVenuesSearchAsync.request)),
    switchMap(action =>
      getObservableVenuesSearch({ action, state$ }).pipe(
        map(adaptGetVenuesSearch),
        map(getVenuesSearchAsync.success),
        catchError(err => of(getVenuesSearchAsync.failure(err))),
        takeUntil(action$.pipe(filter(isActionOf(getVenuesSearchAsync.cancel))))
      )
    )
  )

export const getVenuesSimilarEpic: Epic<
  TRootAction,
  TRootAction,
  NStore.IState
> = (action$, state$) =>
  action$.pipe(
    filter(isActionOf(getVenuesSimilarAsync.request)),
    switchMap(action =>
      getObservableVenuesSimilar({ action, state$ }).pipe(
        map(adaptGetVenuesSimilar),
        map(getVenuesSimilarAsync.success),
        catchError(err => of(getVenuesSimilarAsync.failure(err))),
        takeUntil(
          action$.pipe(filter(isActionOf(getVenuesSimilarAsync.cancel)))
        )
      )
    )
  )

export const getVenuesTrendingEpic: Epic<
  TRootAction,
  TRootAction,
  NStore.IState
> = (action$, state$) =>
  action$.pipe(
    filter(isActionOf(getVenuesTrendingAsync.request)),
    switchMap(action =>
      getObservableVenuesTrending({ action, state$ }).pipe(
        map(adaptGetVenuesTrending),
        map(getVenuesTrendingAsync.success),
        catchError(err => of(getVenuesTrendingAsync.failure(err))),
        takeUntil(
          action$.pipe(filter(isActionOf(getVenuesTrendingAsync.cancel)))
        )
      )
    )
  )

export const getVenuesSuggestCompletionEpic: Epic<
  TRootAction,
  TRootAction,
  NStore.IState
> = (action$, state$) =>
  action$.pipe(
    filter(isActionOf(getVenuesSuggestCompletionAsync.request)),
    switchMap(action =>
      getObservableVenuesSuggestCompletion({ action, state$ }).pipe(
        map(adaptGetVenuesSuggestCompletion),
        map(getVenuesSuggestCompletionAsync.success),
        catchError(err => of(getVenuesSuggestCompletionAsync.failure(err))),
        takeUntil(
          action$.pipe(
            filter(isActionOf(getVenuesSuggestCompletionAsync.cancel))
          )
        )
      )
    )
  )
