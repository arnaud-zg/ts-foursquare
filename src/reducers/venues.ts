import { createReducer } from 'typesafe-actions'
import { NVenues } from '../../types/venues.d'
import { EVenuesAction, TVenuesAction } from '../actions/venues'

export const initialState: NVenues.IState = {
  venues: [],
}

export const venuesReducer = createReducer<NVenues.IState, TVenuesAction>(
  initialState,
  {
    [EVenuesAction.RESOLVE_GET_VENUES_SEARCH]: (state, action) => ({
      ...state,
      venues: action.payload || [],
    }),
  }
)
