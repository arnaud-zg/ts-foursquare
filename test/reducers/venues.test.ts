import { EVenuesAction } from '../../src/actions/venues'
import { venuesReducer } from '../../src/reducers/venues'
import { payload } from '../epics/__mocks__/resolveGetVenuesSearch'

const action = {
  payload: payload.response.venues,
  type: EVenuesAction.RESOLVE_GET_VENUES_SEARCH,
}

describe('reducers/venues', () => {
  test.each`
    scenario                            | state             | action
    ${'should get state after $action'} | ${{ venues: [] }} | ${action}
  `('$scenario with action: $action', ({ state, action }) => {
    expect(venuesReducer(state, action)).toMatchSnapshot()
  })
})
