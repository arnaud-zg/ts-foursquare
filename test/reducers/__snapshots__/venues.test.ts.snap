// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`reducers/venues should get state after {"payload": [Array], "type": "RESOLVE_GET_VENUES_SEARCH"} with action: $action 1`] = `
Object {
  "venues": Array [
    Object {
      "categories": Array [
        Object {
          "icon": Object {
            "prefix": "https://ss3.4sqi.net/img/categories_v2/food/steakhouse_",
            "suffix": ".png",
          },
          "id": "4bf58dd8d48988d1cc941735",
          "name": "Steakhouse",
          "pluralName": "Steakhouses",
          "primary": true,
          "shortName": "Steakhouse",
        },
      ],
      "hasPerk": false,
      "id": "3fd66200f964a5209beb1ee3",
      "location": Object {
        "address": "178 Broadway",
        "cc": "US",
        "city": "Brooklyn",
        "country": "United States",
        "crossStreet": "at Driggs Ave",
        "distance": 10,
        "formattedAddress": Array [
          "178 Broadway (at Driggs Ave)",
          "Brooklyn, NY 11211",
          "United States",
        ],
        "labeledLatLngs": Array [
          Object {
            "label": "display",
            "lat": 40.70995790682886,
            "lng": -73.96229110893742,
          },
        ],
        "lat": 40.70995790682886,
        "lng": -73.96229110893742,
        "postalCode": "11211",
        "state": "NY",
      },
      "name": "Peter Luger Steak House",
      "referralId": "v-1565543515",
      "venuePage": Object {
        "id": "77462637",
      },
    },
  ],
}
`;
